File: packages/gbnf-experiment/data/20260909T154334Z_609edff1/ported_implementation/gbnf/utils/errors/build_error_position.py
from typing import List

from ..char_at import item_at

MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3


def build_error_position(src: str, pos: int) -> List[str]:
    if src == "":
        return ["No input provided"]

    lines = src.split("\n")

    line_idx = 0
    while item_at(lines, line_idx) and pos > len(lines[line_idx]) - 1:
        pos -= len(lines[line_idx])
        line_idx += 1

    lines_to_show: List[str] = []
    start = max(0, line_idx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1))
    for i in range(start, line_idx + 1):
        line = item_at(lines, i)
        lines_to_show.append("" if line is None else line)
    return [
        *lines_to_show,
        " " * max(0, pos) + "^",
    ]

File: proposed-projects/round-trip-experiment/reverse/20260909T015142Z_f8ca89ff/20260910T053604Z_b86749e4/ported_implementation/src/utils/errors/build-error-position.ts
export const MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW = 3;

export const buildErrorPosition = (src: string, pos: number): string[] => {
  if (src === '') {
    return ['No input provided'];
  }

  const lines = src.split('\n');

  let lineIdx = 0;
  // `lines[lineIdx]` is falsy both when out of range and when the line is empty.
  while (lines[lineIdx] && pos > lines[lineIdx].length - 1) {
    pos -= lines[lineIdx].length;
    lineIdx += 1;
  }

  const linesToShow: string[] = [];
  for (
    let i = Math.max(0, lineIdx - (MAXIMUM_NUMBER_OF_ERROR_LINES_TO_SHOW - 1));
    i <= lineIdx;
    i++
  ) {
    // Past the end of the grammar there is nothing to show.
    linesToShow.push(lines[i] ?? '');
  }

  return [...linesToShow, `${' '.repeat(Math.max(0, pos))}^`];
};
